### PR TITLE
feat: added example coverage for HIP-865

### LINKED ISSUE #688: Add coverage example for Transient Storage and Memory Copy Cancun opcodes (HIP-865)
## Problem
Based on [HIP-865](https://hips.hedera.com/hip/hip-865), Transient Storage and Memory Copy Cancun opcodes are now officially enabled on Hedera network. Currently, no contracts or examples existed to showcase this update.

## Solution
Add new contracts and associated tests to expose the support for new Cancun opcodes

## Resources:
- https://eips.ethereum.org/EIPS/eip-1153
- https://eips.ethereum.org/EIPS/eip-5656
- https://soliditylang.org/blog/2024/01/26/solidity-0.8.24-release-announcement/
- https://hips.hedera.com/hip/hip-865

### PR DESCRIPTION
**Pull request by quiet-node**
**Description**:
- upgrade the solidity version to 0.8.24 and specify the evmVersion to Cancun in hardhat.config.js
- adds example coverage for HIP-865 about the new Cancun Opcodes for Transient Storage and Memory Copy.

**Related issue(s)**:

Fixes #688
Related #725

**Notes for reviewer**:
The artifact files of the erc-721 and erc-20 are modified as the solidity version is upgraded.

**Checklist**

- [x] Documented (Code comments, README, etc.)
- [x] Tested (unit, integration, etc.)

### CODE REVIEW

**Comment by quiet-node** [comment]
System Contract DApp failing CIs are reported in this issue https://github.com/hashgraph/hedera-smart-contracts/issues/817

File: contracts/cancun/cancun-opcodes/CancunOpcodes.sol [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+    //      then read `value` from transient storage, using tload, into memory variable, val,
+    //      and finally write `val` to regular storage at `regularSlot`
+    function transientStorage(uint256 value, uint256 transientSlot, uint256 regularSlot) external {
+        // store `value` to `slot` using tstore
+        assembly {
+            tstore(transientSlot, value)
+        }
+
+        // The `value` at `transientSlot` is stored transiently and will be wiped off after the transaction is done processing.
+        // Therefore, in order to pertain the value for the sake of testing, value will be retrieved from the same transientSlot using TLOAD
```
Comment (david-bakin-sl) [inline contracts/cancun/cancun-opcodes/CancunOpcodes.sol:17]: "pertain" is probably not the word you were looking for?

File: test/cancun/cancun-opcodes/transientStorage.js [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+ * limitations under the License.
+ *
+ */
+
+const { expect } = require('chai');
+const { ethers } = require('hardhat');
+const Constants = require('../../constants');
+
+describe('@cancun Transient Storage Test Suite', function () {
+  const VAULE = 7;
```
Comment (david-bakin-sl) [inline test/cancun/cancun-opcodes/transientStorage.js:26]: `VAULE` -> `VALUE`

**Review by david-bakin-sl — APPROVED** [review]
LGTM (minor comments)

File: contracts/cancun/cancun-opcodes/CancunOpcodes.sol (hunk 1 above)
Comment (quiet-node) [reply contracts/cancun/cancun-opcodes/CancunOpcodes.sol:17]: Ah I was confused with the word retain. Great catch thanks!

File: test/cancun/cancun-opcodes/transientStorage.js (hunk 2 above)
Comment (quiet-node) [reply test/cancun/cancun-opcodes/transientStorage.js:26]: Nice catch tahnks!

**Review by david-bakin-sl — APPROVED** [review]
LGTM!

**Pull request merged**